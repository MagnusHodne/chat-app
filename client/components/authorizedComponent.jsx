import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoadingComponent } from "./feedbackComponents";
import { ChatApiContext } from "../chatApiContext";

/**
 * Simple wrapper for components that require the user to be authorized
 * @param user the user object used for checking verification
 * @param {boolean} writeAccess Whether this component needs write access (default is false)
 * @param successComponent
 * @param failComponent
 * @param verificationFunction what function to use for verifying user. By default it uses the one specified in apiMethods
 */
export function AuthorizedComponent({
  user,
  writeAccess = false,
  successComponent = <div>Success</div>,
  failComponent = <div>Fail</div>,
}) {
  const navigate = useNavigate();
  const [verified, setVerified] = useState(false);
  const { verifyUser } = useContext(ChatApiContext);
  useEffect(() => {
    if (!verifyUser(user)) {
      navigate("/startlogin");
    } else {
      setVerified(true);
    }
  }, []);
  if (!verified)
    return <LoadingComponent message={"Redirecting to login..."} />;
  return successComponent;
}
