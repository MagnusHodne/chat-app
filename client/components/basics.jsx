import React from "react";
import { Link, useNavigate } from "react-router-dom";

export function FAIcon({ icon }) {
  return <i className={`icon ${icon}`}></i>;
}

export function Button({ title, icon, to, onClick, className }) {
  if (to) {
    const navigate = useNavigate();
    onClick = () => {
      navigate(to);
    };
  }
  const button = () => (
    <button
      className={`button ${className !== undefined ? className : ""}`}
      onClick={onClick}
    >
      {icon && <FAIcon icon={icon} />}
      {title}
    </button>
  );
  return button();
}
