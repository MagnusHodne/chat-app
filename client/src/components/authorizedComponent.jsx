import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoadingComponent } from "./feedbackComponents";
import { ChatApiContext } from "../chatApiContext";
import { UserContext } from "../userContext";

/**
 * Simple wrapper for components that require the user to be authorized
 * @param user the user object used for checking verification
 * @param {boolean} writeAccess Whether this component needs write access (default is false)
 * @param successComponent
 * @param failComponent
 */
export function AuthorizedComponent({
  writeAccess = false,
  successComponent = <div>Success</div>,
  failComponent = <div>Fail</div>,
}) {
  const navigate = useNavigate();
  const [verified, setVerified] = useState(false);
  const { user } = useContext(UserContext);
  const { verifyUser } = useContext(ChatApiContext);
  useEffect(() => {
    if (!verifyUser(user)) {
      navigate("/startlogin");
    } else {
      setVerified(true);
    }
  }, []);
  if (!verified)
    return <LoadingComponent message={"Redirecting to login..."} />;
  return successComponent;
}
