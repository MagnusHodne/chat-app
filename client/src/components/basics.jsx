import React from "react";
import { Link, useNavigate } from "react-router-dom";

export function FAIcon({ icon, className }) {
  return <i className={`${icon} ${className}`}></i>;
}

export function Button({ title, icon, to, onClick }) {
  if (to) {
    const navigate = useNavigate();
    onClick = () => {
      navigate(to);
    };
  }
  const button = () => (
    <button
      className={`p-2 m-1 rounded-md gap-2 text-thischord-100 bg-thischord-500 flex items-center justify-center hover:bg-thischord-400`}
      onClick={onClick}
    >
      {icon && <i className={`shrink-0 text-thischord-100 ${icon}`} />}
      {title}
    </button>
  );
  return button();
}
