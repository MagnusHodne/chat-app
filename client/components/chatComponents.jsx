import React, { useState } from "react";
import { FAIcon } from "./basics";

export function ChatHeader({ name }) {
  return (
    <div className={"chat-header"}>
      <FAIcon icon={"fa-solid fa-hashtag"} />
      <h3>{name}</h3>
    </div>
  );
}

function ChatMessageActions({ user }) {}

function ChatMessage({ message, user, created, _id, handleDelete }) {
  const date = new Date(created);

  return (
    <div className={"message-container"}>
      <div className={"message-info"}>
        <strong>{user.name}</strong>
        <small className={"message-time"}>
          {`${date.toDateString()} ${date.getHours()}:${date.getMinutes()}`}
        </small>
      </div>
      {message}
      <button className={"message-actions"} onClick={() => handleDelete(_id)}>
        Delete
      </button>
    </div>
  );
}

/**
 *
 * @param messages List of messages
 * @param onNewMessage Function for handling saves
 * @param onDeleteMessage Function for handling deletes
 * @param chatRoom Name of the chat room
 * @returns {JSX.Element}
 * @constructor
 */
export function ChatComponent({
  messages,
  onNewMessage,
  onDeleteMessage,
  chatRoom = "main",
}) {
  const [message, setMessage] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }

  return (
    <div className={"chat-window"}>
      <ChatHeader name={chatRoom} />
      <div className={"scroll"}>
        {messages.map(({ message, user, created, _id }) => (
          <ChatMessage
            message={message}
            user={user}
            created={created}
            key={_id}
            _id={_id}
            handleDelete={onDeleteMessage}
          />
        ))}
      </div>
      <footer>
        <form onSubmit={handleSubmit}>
          <input
            autoFocus={true}
            placeholder={`Message #${chatRoom}`}
            value={message}
            onChange={(e) => {
              setMessage(e.target.value);
            }}
          />
          <button>Send</button>
        </form>
      </footer>
    </div>
  );
}
