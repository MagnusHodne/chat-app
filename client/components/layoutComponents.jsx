import React, { useContext } from "react";
import { Button, FAIcon } from "./basics";
import { ChatApiContext } from "../chatApiContext";

export function Content({ content }) {
  return <div className={"content"}>{content}</div>;
}

function UserActions({ user }) {
  const { verifyUser } = useContext(ChatApiContext);
  if (!verifyUser(user)) {
    return <></>;
  }

  return (
    <>
      <Button to={"/login/endsession"} title={"Log out"} className={"danger"} />
    </>
  );
}
export function Header({ user }) {
  return (
    <header className={"header"}>
      <div className={"title"}>
        <FAIcon icon={"fa-solid fa-gamepad"} />
        <h1>Thischord</h1>
      </div>
      <div>
        <UserActions user={user} />
      </div>
    </header>
  );
}

export function Sidebar({ content }) {
  return <div className={"sidebar"}>{content}</div>;
}
