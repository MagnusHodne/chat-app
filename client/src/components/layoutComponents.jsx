import React, { useContext } from "react";
import { Button, FAIcon } from "./basics";
import { ChatApiContext } from "../chatApiContext";
import { UserContext } from "../userContext";

export function Content({ content }) {
  return <div className={"min-h-0 bg-thischord-600"}>{content}</div>;
}

function UserActions() {
  const { user } = useContext(UserContext);
  const { verifyUser } = useContext(ChatApiContext);
  if (!verifyUser(user)) {
    return <></>;
  }

  return (
    <>
      <Button to={"/login/endsession"} title={"Log out"} className={"danger"} />
    </>
  );
}
export function Header() {
  return (
    <header
      className={
        "col-span-2 flex flex-row items-center justify-between border-b-2 border-solid border-thischord-900 bg-thischord-600 p-4"
      }
    >
      <div className={"flex flex-row items-center gap-3"}>
        <FAIcon icon={"fa-solid fa-gamepad"} />
        <h1>Thischord</h1>
      </div>
      <div>
        <UserActions />
      </div>
    </header>
  );
}

export function Sidebar({ content }) {
  return <div className={"bg-thischord-800 p-2"}>{content}</div>;
}
