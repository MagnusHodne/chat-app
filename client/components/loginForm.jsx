import React from "react";
import { Button } from "./basics";
import { Header } from "./layoutComponents";

export function LoginForm() {
  return (
    <div className={"page-fullsize"}>
      <Header />
      <div className={"content"}>
        <div className={"login-buttons"}>
          <h1>Welcome!</h1>
          <p>Please log in</p>
          <div
            style={{ display: "flex", flexDirection: "column", width: "15em" }}
          >
            <Button
              to={"/login/google"}
              title={"Log in with Google"}
              icon={"fa-brands fa-google"}
            />
            <Button
              to={"/login/azure"}
              title={"Log in with Active Directory"}
              icon={"fa-brands fa-microsoft"}
            />
          </div>
        </div>
      </div>
    </div>
  );
}
