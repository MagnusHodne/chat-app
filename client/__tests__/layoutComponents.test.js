import React from "react";
import ReactDOM from "react-dom";
import { Content, Header, Sidebar } from "../src/components/layoutComponents";
import { BrowserRouter } from "react-router-dom";

let element;
beforeEach(() => {
  element = document.createElement("div");
});

describe("Header", () => {
  it("matches snapshot with no user", () => {
    ReactDOM.render(<Header />, element);
    expect(element).toMatchSnapshot();
  });

  it("matches snapshot with a user", () => {
    ReactDOM.render(
      <BrowserRouter>
        <Header user={{ name: "Test user" }} />
      </BrowserRouter>,
      element
    );
    expect(element).toMatchSnapshot();
  });
});

describe("Sidebar", () => {
  it("matches snapshot", () => {
    ReactDOM.render(<Sidebar />, element);
  });
  expect(element).toMatchSnapshot();
});

describe("Content", () => {
  it("matches snapshot", () => {
    ReactDOM.render(<Content />, element);
  });
  expect(element).toMatchSnapshot();
});
