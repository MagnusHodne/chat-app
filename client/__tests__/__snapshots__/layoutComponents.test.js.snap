// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `undefined`;

exports[` 2`] = `undefined`;

exports[`Header matches snapshot with a user 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="title"
    >
      <i
        class="icon fa-solid fa-gamepad"
      />
      <h1>
        Thischord
      </h1>
    </div>
    <div>
      <button
        class="button danger"
      >
        Log out
      </button>
    </div>
  </header>
</div>
`;

exports[`Header matches snapshot with no user 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="title"
    >
      <i
        class="icon fa-solid fa-gamepad"
      />
      <h1>
        Thischord
      </h1>
    </div>
    <div />
  </header>
</div>
`;
